File: puzzle/corns/solve.py
import copy
from typing import List, Tuple, Optional

# 环形玉米棒子参数
ROWS = 7
COLS = 14

# 示例玉米粒团，可以自行添加更多
CORN_PIECES = [
    [[1, 1, 1], [0, 1, 1]],
    [[1, 1, 1], [0, 1, 1]],
    [[1, 1, 1], [0, 0 ,1]],
    [[1, 1, 0], [1, 1, 1], [0, 1, 1]],
    [[1, 1, 0], [0, 1, 0], [0, 1, 0], [0, 1, 1]],
    [[1, 1, 1], [0, 1, 0], [0, 1, 0], [0, 1, 0]],
    [[1, 1, 1], [0, 0, 1]],
    [[1, 1, 1], [1, 1, 0]],
    [[1, 1, 1], [0, 0, 1], [0, 0, 1]],
    [[1, 1, 1], [1, 0, 1], [0, 0, 1]],
    [[1, 1, 1], [1, 0, 1], [0, 0, 1]],
    [[1, 1, 1], [1, 0, 1], [1, 0, 1]],
    [[1, 1, 0], [1, 1, 1], [0, 0, 1]],
    [[1, 1, 1], [1, 1, 0], [1, 0, 0]],
    [[1, 1, 1], [1, 1, 0], [1, 0, 0], [1, 0, 0]],
    [[0, 1, 1], [1, 1, 1], [0, 1, 0]],
    [[0, 1, 1], [1, 1, 1], [0, 0, 1], [0, 0, 1]],
]

def rotate_piece(piece: List[List[int]]) -> List[List[int]]:
    """顺时针旋转玉米粒团90度"""
    return [list(row) for row in zip(*piece[::-1])]

def get_piece_variants(piece: List[List[int]]) -> List[List[List[int]]]:
    """获取玉米粒团的所有旋转形态（0, 90, 180, 270度）"""
    variants = []
    current = piece
    for _ in range(4):
        if current not in variants:
            variants.append(copy.deepcopy(current))
        current = rotate_piece(current)
    return variants

def can_place(board: List[List[int]], piece: List[List[int]], row: int, col: int) -> bool:
    """判断玉米粒团能否放在指定位置"""
    for i, prow in enumerate(piece):
        for j, val in enumerate(prow):
            if val:
                r = row + i
                c = (col + j) % COLS  # 环形
                if r >= ROWS or board[r][c]:
                    return False
    return True

def place_piece(board: List[List[int]], piece: List[List[int]], row: int, col: int, pid: int):
    """放置玉米粒团到棋盘上"""
    for i, prow in enumerate(piece):
        for j, val in enumerate(prow):
            if val:
                r = row + i
                c = (col + j) % COLS
                board[r][c] = pid

def remove_piece(board: List[List[int]], piece: List[List[int]], row: int, col: int):
    """移除玉米粒团"""
    for i, prow in enumerate(piece):
        for j, val in enumerate(prow):
            if val:
                r = row + i
                c = (col + j) % COLS
                board[r][c] = 0

def find_empty(board: List[List[int]]) -> Optional[Tuple[int, int]]:
    """找到第一个空格"""
    for i in range(ROWS):
        for j in range(COLS):
            if board[i][j] == 0:
                return i, j
    return None

def solve(board: List[List[int]], pieces: List[List[List[int]]], used: List[bool], solution: List[Tuple[int, int, int, int]]) -> Optional[List[Tuple[int, int, int, int]]]:
    """递归搜索解决方案"""
    empty = find_empty(board)
    if empty is None:
        return solution  # 全部填满
    row, col = empty
    for idx, piece in enumerate(pieces):
        if used[idx]:
            continue
        for variant in get_piece_variants(piece):
            for offset in range(COLS):  # 环形每一列都尝试
                if can_place(board, variant, row, (col + offset) % COLS):
                    place_piece(board, variant, row, (col + offset) % COLS, idx + 1)
                    used[idx] = True
                    solution.append((idx, row, (col + offset) % COLS, get_piece_variants(piece).index(variant)))
                    result = solve(board, pieces, used, solution)
                    if result:
                        return result
                    solution.pop()
                    used[idx] = False
                    remove_piece(board, variant, row, (col + offset) % COLS)
    return None

def print_board(board: List[List[int]]):
    for row in board:
        print(' '.join(str(x) if x else '.' for x in row))

def main():
    board = [[0 for _ in range(COLS)] for _ in range(ROWS)]
    used = [False] * len(CORN_PIECES)
    solution = []
    result = solve(board, CORN_PIECES, used, solution)
    if result:
        print("找到解决方案:")
        print_board(board)
        print("放置顺序: (piece_id, row, col, variant_id)")
        print(result)
    else:
        print("无解")

if __name__ == "__main__":
    main()

File: puzzle/corns/solve.cpp
#include <iostream>
#include <vector>
#include <array>
#include <algorithm>
#include <optional>

using namespace std;

// Constants for the board
constexpr int ROWS = 7;
constexpr int COLS = 14;

// Type definitions
using Board = array<array<int, COLS>, ROWS>;
using Piece = vector<vector<int>>;

// The corn pieces
vector<Piece> CORN_PIECES = {
    {{1, 1, 1}, {0, 1, 1}},
    {{1, 1, 1}, {0, 1, 1}},
    {{1, 1, 1}, {0, 0, 1}},
    {{1, 1, 0}, {1, 1, 1}, {0, 1, 1}},
    {{1, 1, 0}, {0, 1, 0}, {0, 1, 0}, {0, 1, 1}},
    {{1, 1, 1}, {0, 1, 0}, {0, 1, 0}, {0, 1, 0}},
    {{1, 1, 1}, {0, 0, 1}},
    {{1, 1, 1}, {1, 1, 0}},
    {{1, 1, 1}, {0, 0, 1}, {0, 0, 1}},
    {{1, 1, 1}, {1, 0, 1}, {0, 0, 1}},
    {{1, 1, 1}, {1, 0, 1}, {0, 0, 1}},
    {{1, 1, 1}, {1, 0, 1}, {1, 0, 1}},
    {{1, 1, 0}, {1, 1, 1}, {0, 0, 1}},
    {{1, 1, 1}, {1, 1, 0}, {1, 0, 0}},
    {{1, 1, 1}, {1, 1, 0}, {1, 0, 0}, {1, 0, 0}},
    {{0, 1, 1}, {1, 1, 1}, {0, 1, 0}},
    {{0, 1, 1}, {1, 1, 1}, {0, 0, 1}, {0, 0, 1}},
};

// Reverse a piece (mirror horizontally and vertically)
Piece reverse_piece(const Piece& p) {
    Piece rev = p;
    reverse(rev.begin(), rev.end());
    for (auto& row : rev) reverse(row.begin(), row.end());
    return rev;
}

// Generate all unique orientations for a piece (original and reversed)
vector<Piece> generate_orientations(const Piece& p) {
    vector<Piece> res;
    res.push_back(p);
    Piece rev = reverse_piece(p);
    if (rev != p) res.push_back(rev);
    return res;
}

// Check if a piece can be placed at (row, col) on the board
bool can_place(const Board& board, const Piece& piece, int row, int col) {
    int h = piece.size();
    int w = piece[0].size();
    for (int i = 0; i < h; ++i) {
        int r = row + i;
        if (r >= ROWS) return false;
        for (int j = 0; j < w; ++j) {
            if (piece[i][j]) {
                int c = (col + j) % COLS;
                if (board[r][c]) return false;
            }
        }
    }
    return true;
}

// Place a piece on the board
void place(Board& board, const Piece& piece, int row, int col, int val) {
    int h = piece.size();
    int w = piece[0].size();
    for (int i = 0; i < h; ++i) {
        int r = row + i;
        for (int j = 0; j < w; ++j) {
            if (piece[i][j]) {
                int c = (col + j) % COLS;
                board[r][c] = val;
            }
        }
    }
}

// Print the board
void print_board(const Board& board) {
    for (int i = 0; i < ROWS; ++i) {
        for (int j = 0; j < COLS; ++j) {
            cout << (board[i][j] ? "#" : ".");
        }
        cout << endl;
    }
}

// Recursive DFS search
bool dfs(Board& board, vector<vector<Piece>>& all_orients, vector<bool>& used, vector<pair<int,int>>& solution) {
    // Find the first empty cell
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (board[r][c] == 0) {
                // Try all unused pieces
                for (int p = 0; p < all_orients.size(); ++p) {
                    if (used[p]) continue;
                    for (const auto& orient : all_orients[p]) {
                        int h = orient.size();
                        int w = orient[0].size();
                        // Try all possible positions for this orientation at (r, c)
                        for (int offset = 0; offset < COLS; ++offset) {
                            int cc = (c - offset + COLS) % COLS;
                            if (can_place(board, orient, r, cc)) {
                                place(board, orient, r, cc, p+1);
                                used[p] = true;
                                solution.push_back({p, cc});
                                if (dfs(board, all_orients, used, solution)) return true;
                                solution.pop_back();
                                used[p] = false;
                                place(board, orient, r, cc, 0);
                            }
                        }
                    }
                }
                return false; // No piece fits here
            }
        }
    }
    // All cells filled
    return true;
}

int main() {
    Board board = {};
    vector<vector<Piece>> all_orients;
    for (const auto& p : CORN_PIECES) {
        all_orients.push_back(generate_orientations(p));
    }
    vector<bool> used(CORN_PIECES.size(), false);
    vector<pair<int,int>> solution;
    if (dfs(board, all_orients, used, solution)) {
        cout << "Solution found:\n";
        print_board(board);
    } else {
        cout << "No solution found.\n";
    }
    return 0;
}